### DISCUSSION TITLE
Support `exclude`, `include.only`, `mask.ok` for packages.

### DISCUSSION THREAD

**Discussion by smingerson** [ropensci/targets#684] [Ideas]
Drake user here, working on my first project using targets. Loving the differences so far.

Currently, you can load packages in as needed for a target by providing `tar_option_set(packages = ...)` in the targets script. The suggested method for more complicated dependency loading is to include it in .Rprofile. Another option I considered was using `tar_hook_before()`. While both of these methods would work fine, I think I should be able to do something like the following. 

```r
tar_option_set(packages = 
  list(purrr = list(exclude = "transpose"),
       "dplyr")
)
#
tar_option_set(packages =
    list(
      tar_pkg(dplyr, exclude = c("intersect", "setdiff", "union")),
      tar_pkg(stats, exclude = c("filter", "lag")),
      "purrr" ) 
)
```

While this complicates things, I don't think it something that would require much, if any, maintenance once implemented since it is just passing extra arguments to `library` or `require`. 

This method would allow only loading packages as needed for each target, while maintaining the ability to be specific about what is needed from each package. Prefixing with the namespace is fine for infrequently called packages, but becomes a hassle for others. 

This is, admittedly, an incredibly minor issue and won't prevent my use of targets.

**Comment by wlandau** [comment]
Neat, I hadn't thought of that. I would prefer to hold off on this though because this is not a common use case and there are other ways to accomplish the same thing in `.Rprofile`. For example, some other users favor the `conflicted` package. `conflicted::prefer()` seems to operate with slightly different and possibly more convenient logic.

**Comment by smingerson** [comment]
Thanks for the suggestion! I have heard of the conflicted package, but hadn't thought to do so here. I'll go with using it.

**Comment by smingerson** [comment]
I'm not sure that using conflicted accomplishes the same goal--ability to only attach a subset of each required package--unless you load all packages prior to the `conflicted::conflict_prefer()` calls, in which case you end up loading all packages. 
```r
# package version: 1.0.4
conflict_prefer <- function (name, winner, losers = NULL, quiet = FALSE) 
{
    stopifnot(is.character(name), length(name) == 1)
    stopifnot(is.character(winner), length(winner) == 1)
    stopifnot(is.null(losers) || is.character(losers))
    if (env_has(prefs, name)) {
        if (!quiet) 
            message("[conflicted] Removing existing preference")
    }
    if (!quiet) {
        full <- style_name(winner, "::", backtick(name))
        if (is.null(losers)) {
            message("[conflicted] Will prefer ", full, 
                " over any other package")
        }
        else {
            alt <- style_name(losers, "::", backtick(name))
            message("[conflicted] Will prefer ", full, 
                " over ", paste(alt, collapse = ", "))
        }
    }
    env_bind(prefs, `:=`(!!name, c(winner, losers)))
# Only registers the conflict if the package is already attached
    if (pkg_attached(winner)) 
        conflicts_register()
    invisible()
}
```

I came across this by putting `conflict_prefer()` calls in my .Rprofile, then restarting the session and calling `tar_load_globals()`. I get the masking warnings from `library()`, which are somewhat expected, but if I print a function body the namespaces are not respecting my conflict preferences.

Here's my attempt at an example. 

```r
library(targets)

tar_dir({
  writeLines("conflicted::conflict_prefer('between', 'data.table')", ".Rprofile")
  tar_script({
    tar_option_set(packages = c("data.table", "dplyr"),
                   library = .libPaths())
    tar_target(namespace,
               environmentName(environment(between))
               )
  })
  tar_make()
  tar_read("namespace")
})

```

**Reply by wlandau** [reply]
Here is a reproducible example without `targets`. If you invoke `conflicted` but then load `dplyr`, `between()` will come from `dplyr`. 

``` r
conflicted::conflict_prefer('between', 'data.table')
#> [conflicted] Will prefer data.table::between over any other package
library(data.table)
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:data.table':
#> 
#>     between, first, last
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
environmentName(environment(between))
#> [1] "dplyr"
```

<sup>Created on 2021-11-10 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

So in your case, rather than loading packages with `tar_option_set(packages = ...)`, I recommend putting it all in your `.Rprofile`: `library()` calls, then `conflicted_prefer()`.

``` r
# .Rprofile
library(data.table)
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:data.table':
#> 
#>     between, first, last
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
conflicted::conflict_prefer('between', 'data.table')
#> [conflicted] Will prefer data.table::between over any other package
environmentName(environment(between))
#> [1] "data.table"
```

<sup>Created on 2021-11-10 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by svraka** [comment]
Just to add my two cents, after experimenting with setting up **conflicted** based in this discussion.

There are a couple of consequences of loading everything in `.Rprofile`:

1. Default packages are loaded after `.Rprofile` is sourced, thus your package loading order will change, unless you work around that by modifying the `defaultPackages` option. The whole exercise revolves around dealing with namespace conflicts, this shouldn't lead to unexpected behaviours because you are likely to declare conflicts with default packages anyway but it might be something to keep in mind.
1. It's inconvenient for utilities like `tar_outdated`, or `tar_visnetwork` as they will be slow if a lot of packages are loaded. If your pipeline is large, computing dependencies will outweigh those few seconds spent loading packages but in my projects, this is not the case, and I want quick responses from these interactiveused utilities.
1. It's impossible to use different packages in different targets. As far as I understand, this point is only relevant if you run a target with custom packages outside the main session of your pipeline with **future**, or on a cluster, as otherwise all packages will be loaded eventually (during which your namespace conflicts are not guaranteed to resolve as expected).

Anyway, with these caveats it is possible to use **conflicted** alongside **targets**' package options, as **conflicted** sets up hooks on package *attachment* but preferences can be declared by just *loading* it. Thus if you declare preferences in `.Rprofile` with `conflicted::conflict_prefer(...)` and load **conflicted** after all other packages in the `packages` option for **targets** (`tar_option_set(packages = c("package1", "package2", "conflicted"))`), it should work. You need to be careful, if you want to change loaded packages for a specific target, you need to make sure **conflicted** stays the last target (e.g. `packages = c("some", "new", "packages", tar_option_get("packages"))`).

So, I suppose there's no universal solution when setting up **conflicted** with **targets** and you need to consider each project and use case.
